### ISSUE TITLE
[DOC]: Is it required that groupby key expressions are elementwise?

### ISSUE THREAD

**Issue by wence-** [pola-rs/polars#20152]
### Description

Consider:

```
import polars as pl

df = pl.DataFrame({"key": [1, 2, 1], "value": [1, 2, 3]})

result = df.group_by(pl.col("key").sort()).agg(pl.col("value"))
```

This (currently) runs and produces:
```
shape: (2, 2)
┌─────┬───────────┐
│ key ┆ value     │
│ --- ┆ ---       │
│ i64 ┆ list[i64] │
╞═════╪═══════════╡
│ 1   ┆ [1, 2]    │
│ 2   ┆ [3]       │
└─────┴───────────┘
```

However, it's kind of nonsensical, because the sort of the key expression destroys any relationship between the keys and the values of the input dataframe.

So, question, what types of expressions _should_ be valid as groupby keys (and, corollary, should polars enforce the validity at plan lowering time?).

Some ideas:

1. Only elementwise expressions of column expressions (rules out keys that consist only of literals)
2. Any expression that is (broadcastable to) the same length as the input frame
3. Something else here

2. Appears to be the current status quo.

Note there were some discussions (and subsequent tightening of the rules) for arguably the same kind of thing for joins: https://github.com/pola-rs/polars/issues/17184 (see also https://github.com/pola-rs/polars/issues/17895)

**Comment by rodrigogiraoserrao** [comment]
I don't think I understand fully what you're suggesting. Key expressions don't need to be elementwise to be helpful.
Here's an example I concocted where I have 3 groups where athletes faced off.
Now, I want the winning athletes to face off against each other and the losing athletes to face off against each other:

```py
df = pl.DataFrame({
    "time": [10.1, 10.3, 10.08, 11, 10.5, 10.12],
    "group": ["a", "a", "b", "b", "c", "c"],
    "name": ["A", "B", "C", "D", "E", "F"],
})
print(
    df.group_by(pl.col("time").rank().over("group").alias("rank")).agg(pl.col("name"))
)
```
```
shape: (2, 2)
┌──────┬─────────────────┐
│ rank ┆ name            │
│ ---  ┆ ---             │
│ f64  ┆ list[str]       │
╞══════╪═════════════════╡
│ 2.0  ┆ ["B", "D", "E"] │
│ 1.0  ┆ ["A", "C", "F"] │
└──────┴─────────────────┘
```

I can't see how Polars would look at an expression and determine whether it's nonsensical or not, since the result you got matches exactly what you told it to do...

Of course, I might be missing something here, in which case I'd appreciate you point it out :D

**Comment by wence-** [comment]
Thanks, that is a good counterexample to "elementwise". However, your expression does morally act "elementwise" in that it keeps the association between the new column and the existing columns. So it "makes sense". But if `.rank.over()` didn't guarantee giving things back in the original order, then your groupby wouldn't make sense.

**Comment by avimallu** [comment]
Here's a contrived example I thought of where this action is deliberate:
> [operation on] key expression destroys any relationship between the keys and the values of the input dataframe.

I want to randomly assign `elements` into 3 groups - a sort of random sampling, if you will.

```py
import polars as pl
elements = ["a", "b", "c", "d", "e", "f", "g", "h", "i"]
(
   pl.DataFrame({"elements": elements})
   .group_by(pl.int_range(0, pl.len()).floordiv(3).shuffle())
   .agg(pl.col("elements"))
)
```

Since `elements` can be anything, this operation would make sense if I want to apply some convoluted "sample from population" operation. Contrived, yes, but here we deliberately don't care about retaining the relationship between the keys and the value of the input DataFrame.

(This doesn't hold if I misunderstood what "elementwise" means to you.)

**Comment by wence-** [comment]
No, that would be disallowed. I think in that scenario you would do:

```
df.with_columns(
    pl.int_range(0, pl.len()).floordiv(3).shuffle().alias("what")
).group_by("what").agg(pl.col("elements"))
```
